manual/manual_compiler.rst

```diff
@@ -817,3 +817,198 @@ This measurement partitioning is built into the :py:meth:`get_operator_expectati
+is (ideally) the same as if we measured the two qubits first and then applied a
+classical controlled-NOT on the measurement bits. If the gate were a CY instead
+of a CX the effect would be identical: the only difference is the insertion of a
+diagonal operator, whose effect is unmeasurable.
+
+This simplification is effected by the :py:meth:`SimplifyMeasured` pass.
+
+Let's illustrate this with a Bell circuit:
+
+.. jupyter-execute::
+
+    from pytket.passes import SimplifyMeasured
+
+    c = Circuit(2).H(0).CX(0, 1).measure_all()
+    c.qubit_discard_all()
+    SimplifyMeasured().apply(c)
+    print(c.get_commands())
+
+The CX gate has disappeared, replaced with a classical transform acting on the
+bits after the measurement.
+
+Contextual optimisation in practice
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The above three passes are combined in the :py:meth:`ContextSimp` pass, which
+also performs a final :py:meth:`RemoveRedundancies`. Normally, before running a
+circuit on a device you will want to apply this pass (after using
+:py:meth:`Circuit.qubit_create_all` and :py:meth:`Circuit.qubit_discard_all` to
+enable the simplifications).
+
+However, most backends cannot process the classical operations that may be
+introduced by :py:meth:`SimplifyMeasured` or (possibly)
+:py:meth:`SimplifyInitial`. So pytket provides a method
+:py:meth:`separate_classical` to separate the classical postprocessing circuit
+from the main circuit to be run on the device. This postprocessing circuit is
+then passed as the ``ppcirc`` argument to :py:meth:`BackendResult.get_counts` or
+:py:meth:`BackendResult.get_shots`, in order to obtain the postprocessed
+results.
+
+Much of the above is wrapped up in the utility method
+:py:meth:`prepare_circuit`. This takes a circuit, applies
+:py:meth:`Circuit.qubit_create_all` and :py:meth:`Circuit.qubit_discard_all`,
+runs the full :py:meth:`ContextSimp` pass, and then separates the result into
+the main circuit and the postprocessing circuit, returning both.
+
+Thus a typical usage would look something like this:
+
+.. jupyter-execute::
+
+    from pytket.utils import prepare_circuit
+    from pytket.extensions.qiskit import AerBackend
+
+    b = AerBackend()
+    c = Circuit(2).H(0).CX(0, 1)
+    c.measure_all()
+    c0, ppcirc = prepare_circuit(c)
+    b.compile_circuit(c0)
+    h = b.process_circuit(c0, n_shots=10)
+    r = b.get_result(h)
+    shots = r.get_shots(ppcirc=ppcirc)
+    print(shots)
+
+This is a toy example, but illustrates the principle. The actual circuit sent to
+the backend consisted only of a Hadamard gate on qubit 0 and a single
+measurement to bit 0. The classical postprocessing circuit set bit 1 to zero and
+then executed a controlled-NOT from bit 0 to bit 1. These details are hidden
+from us (unless we inspect the circuits), and what we end up with is a shots
+table that is indistinguishable from running the original circuit but with less
+noise.
```

Comment:
Showing that this has less noise would be a nice optional thing to demonstrate, i.e. give the AerBackend a noise model and show that the result from this always looks like a perfect Bell state measurement. Would probably be better in the notebook than the manual, and I don't think it adds a lot on top of what's already here, but it would help to drive home that these can be actually effective in reducing observed noise. I'll leave it up to you if you want to add this or not